source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2021 Shiyu Liu
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_wrap_vector_synonym.h"
+#include "source/fuzz/transformation_composite_construct.h"
+#include "source/opt/function.h"
+#include "source/opt/module.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/opt/instruction.h"
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "test/fuzz/fuzz_test_util.h"
+#include <algorithm>
+#include "source/fuzz/fact_manager/fact_manager.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(uint32_t instruction_id, uint32_t vec_id1, uint32_t vec_id2,
+                                                                 uint32_t vec_id3,  uint32_t vec_type_id, uint32_t pos,
+                                                                 std::vector<uint32_t>& vec1_elements, std::vector<uint32_t>& vec2_elements) {
+    message_.set_instruction_id(instruction_id);
+    message_.set_vec_id1(vec_id1);
+    message_.set_vec_id2(vec_id2);
+    message_.set_vec_id3(vec_id3);
+    message_.set_vec_type_id(vec_type_id);
+    message_.set_scalar_position(pos);
+    for(auto id : vec1_elements) {
+      message_.add_vec1_elements(id);
+    }
+    for(auto id : vec2_elements) {
+      message_.add_vec2_elements(id);
+    }
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+      opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+    auto valid_arithmetic_types = std::unordered_set<SpvOp> {SpvOpIAdd, SpvOpISub, SpvOpIMul, SpvOpFAdd, SpvOpFSub, SpvOpFMul};
+    auto instruction = ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+    auto vector_type = ir_context->get_def_use_mgr()->GetDef(message_.vec_type_id());
```

Comment:
You don't actually need to pass the result id of a vector type into the transformation. Instead, it's better to check whether the module has a suitable type in `IsApplicable` (e.g. by using `fuzzerutil::MaybeGetVectorType`) and create a suitable type in the fuzzer pass (`FuzzerPass::FindOrCreateVectorType` method).

Reply:
Additionally, it might be easier to work with the type if you use the type manager instead (`ir_context->get_type_mgr()->GetId(...)`).